Improve documentation (#18)
supabase-community/chatgpt-your-files: README.md

    );
    ```

>     _Note: Since the video was published, `on delete cascade` was
>     added here so that the lifecycle of `document_sections` is tied
>     to their respective document. We leave the lifecycle of the
>     objects in storage as an exercise to the reader._

Reviewer: Just a minor reword:

```suggestion
    _Note: Since the video was published, `on delete cascade` was
    added as a new migration so that the lifecycle of `document_sections`
    is tied to their respective document._
```

I removed the `exercise to the reader` part just to prevent confusion. I'm okay with adding an `on delete cascade` on that foreign key too as long as we add a note along with it.

Let's also add the SQL code for the `document_sections` `on delete cascade` migration under this note so that users that are going through this guide step by step can copy and paste that logic right away.

    Feel free to adjust these according to your needs. A larger batch size will require a longer timeout per request, since each invocation will have more embeddings to generate. A smaller batch size can use a lower timeout.

>     <details>
>     <summary><i>Note: Lifecycle of triggered edge functions</i></summary>
>     If the triggered edge function fails, you will end up with
>     document sections missing embeddings. In the case of this
>     tutorial, we can `supabase db reset`. In a more production
>     setting, some potential options (although beyond the scope of this tutorial) are:
> 
>       * Add another function that can be triggered manually which checks for `document_sections` with missing embeddings and invokes the `/embed` edge function for them.
>       * Create a [scheduled function](https://supabase.com/docs/guides/functions/schedule-functions) that periodically checks for `document_sections` with missing embeddings and re-generates them. We would likely need to add a locking mechanism (ie. via another column) to prevent the scheduled function from conflicting with the normal `embed` trigger.
>     </details>

Reviewer: ```suggestion
    <details>
    <summary><i>Note: Lifecycle of triggered edge functions</i></summary>
    If the triggered edge function fails, you will end up with
    document sections missing embeddings. During development,
    we can run `supabase db reset` to reset the database. In production,
    some potential options are:

      - Add another function that can be triggered manually which checks for `document_sections` with missing embeddings and invokes the `/embed` edge function for them.
      - Create a [scheduled function](https://supabase.com/docs/guides/functions/schedule-functions) that periodically checks for `document_sections` with missing embeddings and re-generates them. We would likely need to add a locking mechanism (ie. via another column) to prevent the scheduled function from conflicting with the normal `embed` trigger.
    </details>
```

    );
    ```

>     _Note: Transformers.js requires an ONNX model. This means
>     that the model you specify in the pipeline must have an `.onnx`
>     artefact otherwise you will face this helpful error message
>     `Could not locate file [...] xxx.onnx`. We go into more details
>     in [this other tutorial](https://www.youtube.com/watch?v=QdDoFfkVkcw&t=3825s). To convert your models to ONNX, see
>     the documentation [here](https://huggingface.co/docs/transformers.js/en/custom_usage#convert-your-models-to-onnx)._
> 

Reviewer: ```suggestion
    _Note: Transformers.js requires models to exist in the ONNX format. Specifically
    the Hugging Face model you specify in the pipeline must have an `.onnx` file
    under the `./onnx` folder, otherwise you will see the error
    `Could not locate file [...] xxx.onnx`. Check out
    [this explanation](https://www.youtube.com/watch?v=QdDoFfkVkcw&t=3825s) for more details.
    To convert an existing model (eg. PyTorch, Tensorflow, etc) to ONNX, see
    the [custom usage documentation](https://huggingface.co/docs/transformers.js/en/custom_usage#convert-your-models-to-onnx)._
```

#### Create `chat` Edge Function

> _Note: In the following steps, we show how to integrate with the
>  models provided by OpenAI. There are alternatives though.
> Whichever provider you choose, you can reuse the provided code
> as long as the provider offers an OpenAI-compatible API.
> Such providers are [together.ai](https://docs.together.ai/docs/openai-api-compatibility#nodejs), 
> [fireworks.ai](https://readme.fireworks.ai/docs/openai-compatibility), 
> [endpoints.anyscale.com](https://docs.endpoints.anyscale.com/examples/work-with-openai/) 
> or a local model served with [Ollama](https://github.com/ollama/ollama/blob/main/docs/openai.md#openai-javascript-library). 
> You can find more providers [here](https://sdk.vercel.ai/docs/guides).
> We will add the code for Ollama using Notes, 
> and we leave the rest as an exercise for the reader._
> 

Reviewer: ```suggestion
_Note: In this tutorial we use models provided by OpenAI to implement the chat logic.
However since making this tutorial, many new LLM providers exist, such as:
- [together.ai](https://docs.together.ai/docs/openai-api-compatibility#nodejs)
- [fireworks.ai](https://readme.fireworks.ai/docs/openai-compatibility)
- [endpoints.anyscale.com](https://docs.endpoints.anyscale.com/examples/work-with-openai/) 
- [local models served with Ollama](https://github.com/ollama/ollama/blob/main/docs/openai.md#openai-javascript-library)

Whichever provider you choose, you can reuse the code below (that uses the OpenAI lib) as long as they offer an OpenAI-compatible API _(all of providers listed above do)_. We'll discuss how to do this in each step using Ollama, but the same logic applies to all providers
```

    _Note: we must also return CORS headers here (or anywhere else we send a response)._

    <details>
    <summary><i>Note: Ollama support</i></summary>
    Change the model to a model you're serving locally, for example:

    ```diff
    -     model: 'gpt-3.5-turbo-0613',
    +     model: 'dolphin-mistral',
    ```

    Also, Ollama supports CORS by default from localhost.
    If you were calling your Ollama model from a remote host,

>     you would need to allow those origins specifically as per [these docs](https://github.com/ollama/ollama/blob/main/docs/faq.md#how-can-i-allow-additional-web-origins-to-access-ollama).

Reviewer: CORS should apply here since we are connecting via the backend, right?

Author: I find my formulation confusing now that a week has gone by... And I will remove those Ollama-specific details which I found useful to make my code work, but in hindsight are not specific to this tutorial.

More details:
By this I meant that when I dug into Ollama, I saw they had custom logic to allow origins for `localhost` and `127.0.0.1` and `0.0.0.0`: https://github.com/ollama/ollama/blob/e5596c194481a2a2d7483757ad8f6469a8367f8f/server/routes.go#L903-L907 and https://github.com/ollama/ollama/blob/e5596c194481a2a2d7483757ad8f6469a8367f8f/server/routes.go#L931-L938
So this means that in case you're connecting to `localhost:11434` but e.g. serving models on `0.0.0.0`, requests would not be blocked due to CORS checks, even though you could expect this behaviour because they are different hosts https://stackoverflow.com/questions/5256251/are-127-0-0-1-and-localhost-considered-as-two-different-domains-by-browsers.

Reviewer: Thanks @louisguitton - I ended up removing the note on CORS entirely since CORS would only apply if you were to connect directly to Ollama's API from the browser (ie. via `fetch()`). Since this guide performs LLM logic in the backend via edge functions, CORS shouldn't apply and I'd prefer to not to encourage modifying CORS origins unless needed (from a security perspective).

If you think I've missed/overlooked something here, please let me know!
